File: WordPress/Classes/ViewRelated/Post/Categories/WPCategoryTree.m
#import "WPCategoryTree.h"

@implementation WPCategoryTree

- (id)initWithParent:(PostCategory *)parent
{
    if (self = [super init]) {
        self.parent = parent;
        self.children = [NSMutableArray array];
    }

    return self;
}

- (void)getChildrenFromObjects:(NSArray *)collection
{
    NSUInteger count = [collection count];

    for (NSUInteger i = 0; i < count; i++) {
        PostCategory *category = [collection objectAtIndex:i];

        // self.parent can be nil, so compare int values to avoid badness
        if ([category.parentID intValue] == [self.parent.categoryID intValue]) {
            WPCategoryTree *child = [[WPCategoryTree alloc] initWithParent:category];
            [child getChildrenFromObjects:collection];
            [self.children addObject:child];
        }
    }
}

- (NSArray *)getAllObjects
{
    NSMutableArray *allObjects = [NSMutableArray array];
    NSUInteger count = [self.children count];

    if (self.parent) {
        [allObjects addObject:self.parent];
    }

    for (NSUInteger i = 0; i < count; i++) {
        [allObjects addObjectsFromArray:[[self.children objectAtIndex:i] getAllObjects]];
    }

    return allObjects;
}

@end

File: WordPress/Classes/ViewRelated/Post/Categories/WPCategoryTree.swift
import Foundation

class WPCategoryTree: NSObject {
    var parent: PostCategory?
    var children = [WPCategoryTree]()

    @objc init(parent: PostCategory?) {
        self.parent = parent
    }

    @objc func getChildrenFromObjects(_ collection: [Any]) {
        collection.forEach {
            guard let category = $0 as? PostCategory else {
                return
            }

            if isParentChild(category: category, parent: parent) {
                let child  = WPCategoryTree(parent: category)
                child.getChildrenFromObjects(collection)
                children.append(child)
            }
        }
    }

    @objc func getAllObjects() -> [PostCategory] {
        var allObjects = [PostCategory]()
        if let parent = parent {
            allObjects.append(parent)
        }

        children.forEach {
            allObjects.append(contentsOf: $0.getAllObjects())
        }
        return allObjects
    }

    private func isParentChild(category: PostCategory, parent: PostCategory?) -> Bool {
        guard let parent = parent else {
            return category.parentID == 0
        }

        return category.parentID == parent.categoryID
    }
}
